### PR TITLE
Add Sigmoid (cheby), Static Methods (stack, cat)

### LINKED ISSUE #137: Implement Chebyshev Polynomial Approximation for Sigmoid
The doc string mentions 3 algorithms, but only 2 seem to be implemented and tested. Add the Chebyshev algo. 

https://github.com/OpenMined/SyMPC/blob/7561dec407b230767b9dc517553ea77481d891e9/src/sympc/approximations/sigmoid.py#L15-L18

### PR DESCRIPTION
**Pull request by Syzygianinfern0**
## Description
Fixes #137 
Added the Chebyshev Approx for Sigmoid (was present in the docstring but the code was absent).

## How has this been tested?
- Added test case

## Reference [ [```Paper```](http://www.nnw.cz/doi/2012/NNW.2012.22.023.pdf) ]

```bibtex
@article{article,
    author = {Vlcek, Miroslav},
    year = {2012},
    month = {01},
    pages = {387-393},
    title = {Chebyshev polynomial approximation for activation sigmoid function},
    volume = {22},
    journal = {Neural Network World},
    doi = {10.14311/NNW.2012.22.023}
}
```

image

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/approximations/sigmoid.py [hunk 1]
```diff
@@ -41,3 +44,30 @@ def sigmoid(tensor: MPCTensor, method: str = "exp") -> "MPCTensor":
             result += (tensor ** d) * weights[i + 1]
 
         return result
+
+    elif method == "chebyshev":
+        # Make sure the elements are all positive
+        _sign = sign(tensor)
+        tensor_8 = tensor * _sign / 8
```
Comment (LaRiffle) [inline src/sympc/approximations/sigmoid.py:51]: maybe detail why you divide by 8?

File: src/sympc/approximations/sigmoid.py [hunk 2]
```diff
@@ -19,6 +19,9 @@ def sigmoid(tensor: MPCTensor, method: str = "exp") -> "MPCTensor":
 
     Returns:
         tensor: the calulated sigmoid value
+
+    Raises:
+        ValueError: if the given method is not supported
```
Comment (LaRiffle) [inline src/sympc/approximations/sigmoid.py:24]: you can put the reference of the paper in the docsting as well :)

File: src/sympc/approximations/sigmoid.py [hunk 3]
```diff
@@ -41,3 +44,30 @@ def sigmoid(tensor: MPCTensor, method: str = "exp") -> "MPCTensor":
...
+
+        polynomial = 0 * tensor_8
+        for mu in range(p + 1):
+            a_n = factorial(mu + q) / (factorial(mu) * factorial(q))
+            T_n_w = ((1 - tensor_8) / 2) ** mu
+            polynomial += a_n * T_n_w * _sign
+
+        result = scaler * polynomial
+
+        return ((1 - _sign) * (1 + result) + (1 + _sign) * (result)) / 2
```
Comment (LaRiffle) [inline src/sympc/approximations/sigmoid.py:70]: so when it's negative we actually return `(1 + result)`, why is so ?

What is the impact if tensor_8 was not positive?

File: src/sympc/approximations/sigmoid.py (hunk 1 above)
Comment (LaRiffle) [reply src/sympc/approximations/sigmoid.py:51]: and rename tensor_8 accordingly

**Review by LaRiffle — CHANGES_REQUESTED** [review]
Excellent! Just a few questions/suggestions :)

File: src/sympc/approximations/sigmoid.py (hunk 1 above)
Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py:51]: > maybe detail why you divide by 8?

That's a part of way the approximation works. Check out the definition of the sigmoid function used in the image in the description.

image


> rename tensor_8 accordingly

Any ideas for an appropriate name? I felt that `tensor_8` was fine :sweat_smile:

File: src/sympc/approximations/sigmoid.py (hunk 2 above)
Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py:24]: Done!

File: src/sympc/approximations/sigmoid.py (hunk 3 above)
Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py:70]: ### Please Click the Image Below
[](https://www.desmos.com/calculator/c5wcl6dzqy)

File: src/sympc/approximations/sigmoid.py (hunk 1 above)
Comment (LaRiffle) [reply src/sympc/approximations/sigmoid.py:52]: I don't understand in the taylor formula where it is said that `w` should be tensor/8 ? This is just scaling right?

File: src/sympc/approximations/sigmoid.py (hunk 3 above)
Comment (LaRiffle) [reply src/sympc/approximations/sigmoid.py:71]: I feel stupid - shouldn't it `(1 - result)` then, when sign is negative?

Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py:71]: I actually multiplied the sign earlier, hence leading to a double sign inversion.

https://github.com/OpenMined/SyMPC/pull/136/files#diff-4e30e76758ec501224f7729e0b211d42a81a73e390f68f83311c16f0c625bc75R67

I will remove both of them.

Comment (LaRiffle) [reply src/sympc/approximations/sigmoid.py]: ah great!!

File: src/sympc/approximations/sigmoid.py (hunk 1 above)
Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py]: Putting down the discussion here as well. 
- Scaling issue
  - `R` = Regular sigmoid function (that we all know and love) = `1/(1+e^-x)`
  - `S` = Standard sigmoid function (defined in the paper) = `1/(1+e^-8x)`
  - Hence, `R(x)` = `S(x/8)`

- Renamed to `positive_tensor_rescaled`

**Review by LaRiffle — APPROVED** [review]
🙌

File: src/sympc/approximations/sigmoid.py [hunk 4]
```diff
@@ -41,3 +44,31 @@ def sigmoid(tensor: MPCTensor, method: str = "exp") -> "MPCTensor":
...
+    elif method == "chebyshev":
+        # Reference: http://www.nnw.cz/doi/2012/NNW.2012.22.023.pdf
+        # Make sure the elements are all positive
+        _sign = sign(tensor)
+        positive_tensor_rescaled = tensor * _sign / 8
+        p = 8
+        q = 8
+        scaler = ((1 + positive_tensor_rescaled) / 2) ** (q + 1)
+
+        def factorial(n):
```
Comment (gmuraru) [inline src/sympc/approximations/sigmoid.py:57]: MAYBE: Cache?

File: src/sympc/approximations/sigmoid.py [hunk 5]
```diff
@@ -41,3 +44,31 @@ def sigmoid(tensor: MPCTensor, method: str = "exp") -> "MPCTensor":
             result += (tensor ** d) * weights[i + 1]
 
         return result
+
+    elif method == "chebyshev":
```
Comment (gmuraru) [inline src/sympc/approximations/sigmoid.py:48]: Q: There was also this implementation in the CrypTen repo - maybe it was double-checking this implementation with theirs?

File: src/sympc/approximations/sigmoid.py (hunk 4 above)
Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py:57]: Cache the factorial? Considering integers, the computation time should be negligible already.

File: src/sympc/approximations/sigmoid.py (hunk 5 above)
Comment (Syzygianinfern0) [reply src/sympc/approximations/sigmoid.py:48]: https://github.com/facebookresearch/CrypTen/blob/eafda069e7fd19c91b6b08b678342019a95862f3/crypten/common/approximations.py#L326

Seems to be differently implemented :thinking:

**Review by gmuraru — CHANGES_REQUESTED** [review]
Really great changes! Could you also verify if this method is faster than the one in CrypTen - I am thinking we should go with this if this has a better time or accuracy than CrypTen.

**Comment by gmuraru** [comment]
> Really great changes! Could you also verify if this method is faster than the one in CrypTen - I am thinking we should go with this if this has a better time or accuracy than CrypTen, if not, maybe we could add the implementation they have? (if that is a big overhead than we can stick with what we have here)

Don't forget about the comparison! Adding this only to make sure this is not merged before we get a proper comparison! maybe what CrypTen has is better and we would want that to be implemented!

**Comment by gmuraru** [comment]
@Syzygianinfern0 any update on this?  did you tried the CrypTen implementation to check if the performance is better?

File: src/sympc/approximations/tanh.py [hunk 6]
```diff
@@ -31,5 +40,92 @@ def tanh(tensor, method="sigmoid"):
...
+        tanh_polys_flipped = tanh_polys.unsqueeze(dim=-1).T.squeeze(dim=0)
+        out = tanh_polys_flipped.matmul(coeffs)
+
+        # truncate outside [-maxval, maxval]
+        return hardtanh(out)
     else:
         raise ValueError(f"Invalid method {method} given for tanh function")
+
+
+def hardtanh(tensor, min_value=-1, max_value=1):
```
Comment (gmuraru) [inline src/sympc/approximations/tanh.py:57]: MAYBE: Add typehints to the function

File: src/sympc/approximations/tanh.py [hunk 7]
```diff
@@ -31,5 +40,92 @@ def tanh(tensor, method="sigmoid"):
...
+    Returns:
+        MPCTensor: calculated MPCTensor
+    """
+    intermediate = [relu(tensor - min_value), relu(tensor - max_value)]
+    intermediate = intermediate[0] - intermediate[1]
+    return intermediate + min_value
+
+
+@functools.lru_cache(maxsize=10)
+def chebyshev_series(func, width, terms):
```
Comment (gmuraru) [inline src/sympc/approximations/tanh.py:79]: MAYBE: Type hints

File: src/sympc/approximations/tanh.py [hunk 8]
```diff
@@ -31,5 +40,92 @@ def tanh(tensor, method="sigmoid"):
...
+    """
+    n_range = torch.arange(start=0, end=terms).float()
+    x = width * torch.cos((n_range + 0.5) * np.pi / terms)
+    y = func(x)
+    cos_term = torch.cos(torch.ger(n_range, n_range + 0.5) * np.pi / terms)
+    coeffs = (2 / terms) * torch.sum(y * cos_term, axis=1)
+    return coeffs
+
+
+def _chebyshev_polynomials(tensor, terms):
```
Comment (gmuraru) [inline src/sympc/approximations/tanh.py:103]: MAYBE: Also here

File: src/sympc/tensor/static.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+# third party
+import torch
+
+from sympc.session import get_session
+from sympc.tensor.share_tensor import ShareTensor
+from sympc.utils import parallel_execution
+
+
+def stack(tensors, dim=0):
```
Comment (gmuraru) [inline src/sympc/tensor/static.py:16]: MAYBE: Also here

File: src/sympc/tensor/static.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        ShareTensorPointer: Respective shares after stacking
+    """
+    session = get_session(session_uuid_str)
+    result = ShareTensor(session_uuid=UUID(session_uuid_str), config=session.config)
+
+    result.tensor = torch.stack([share.tensor for share in shares])
+    return result
+
+
+def cat(tensors, dim=0):
```
Comment (gmuraru) [inline src/sympc/tensor/static.py:64]: MAYBE: Type hints

File: src/sympc/tensor/static.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+    expected_shape = torch.cat(
+        [torch.empty(each_tensor.shape) for each_tensor in tensors], dim=dim
+    ).shape
+    result = MPCTensor(shares=stack_shares, session=session, shape=expected_shape)
+
+    return result
+
+
+def cat_share_tensor(session_uuid_str, *shares):
```
Comment (gmuraru) [inline src/sympc/tensor/static.py:94]: MAYBE: Type hints

**Review by gmuraru — CHANGES_REQUESTED** [review]

File: setup.cfg [hunk 12]
```diff
@@ -75,7 +75,7 @@ extras = True
 # e.g. --cov-report html (or xml) for html/xml output or --junitxml junit.xml
 # in order to write a coverage file that can be read by Jenkins.
 addopts =
-    --cov sympc --cov-report term-missing
+;    --cov sympc --cov-report term-missing
```
Comment (gmuraru) [inline setup.cfg:78]: Add this back

File: tests/sympc/approximations/sigmoid_test.py [hunk 13]
```diff
@@ -8,7 +8,11 @@
 from sympc.tensor.mpc_tensor import MPCTensor
 
 
-@pytest.mark.parametrize("method", ["maclaurin", "exp"])
+@pytest.mark.parametrize(
+    # "method", ["maclaurin", "exp", "chebyshev", "chebyshev-aliter"]
```
Comment (gmuraru) [inline tests/sympc/approximations/sigmoid_test.py:12]: Q: This needs to get back in, right?

File: setup.cfg (hunk 12 above)
Comment (Syzygianinfern0) [reply setup.cfg:78]: Yep 😅 
I usually remove that as it gets in the way of PyCharm's pytest plugin for some reason 😛

File: tests/sympc/approximations/sigmoid_test.py (hunk 13 above)
Comment (Syzygianinfern0) [reply tests/sympc/approximations/sigmoid_test.py:12]: Ah yes

File: setup.cfg (hunk 12 above)
Comment (Syzygianinfern0) [inline setup.cfg:78]:
```suggestion
    --cov sympc --cov-report term-missing
```

File: tests/sympc/approximations/sigmoid_test.py [hunk 14]
```diff
@@ -8,7 +8,11 @@
...
+    "method",
+    ["chebyshev"],
+)
```
Comment (Syzygianinfern0) [inline tests/sympc/approximations/sigmoid_test.py:15]:
```suggestion
@pytest.mark.parametrize(
    "method", ["maclaurin", "exp", "chebyshev", "chebyshev-aliter"]
)
```

**Review by Syzygianinfern0 — COMMENTED** [review]
Lazy to open my IDE 😅

**Review by gmuraru — APPROVED** [review]
LGTM!

File: tests/sympc/tensor/static_test.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    y_secret = torch.Tensor([-4, 3, -2, 1, 0.0])
+    secret_stacked = torch.stack([x_secret, y_secret])
+
+    session = Session(parties=clients)
+    SessionManager.setup_mpc(session)
+    x = MPCTensor(secret=x_secret, session=session)
+    y = MPCTensor(secret=y_secret, session=session)
+    stacked = stack([x, y])
+
+    assert torch.equal(secret_stacked, stacked.reconstruct())
```
Comment (LaRiffle) [inline tests/sympc/tensor/static_test.py:24]: You have exact equality? :O

File: src/sympc/tensor/static.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
+"""Module level "static" functions.
+
+We use this to support torch.method(Tensor)
```
[the remark this answers is not in the document]
Comment (Syzygianinfern0) [reply src/sympc/tensor/static.py:3]: Yes, will change that 👍

File: tests/sympc/tensor/static_test.py (hunk 15 above)
Comment (Syzygianinfern0) [reply tests/sympc/tensor/static_test.py:24]: Yep we do 😛

File: src/sympc/approximations/tanh.py [hunk 17]
```diff
@@ -31,5 +43,94 @@ def tanh(tensor, method="sigmoid"):
...
+
+    elif method == "chebyshev":
+        terms = 10  # Higher terms gives slower but more accurate results
+        coeffs = chebyshev_series(torch.tanh, 1, terms)[1::2]
+        tanh_polys = _chebyshev_polynomials(tensor, terms)
+        tanh_polys_flipped = tanh_polys.unsqueeze(dim=-1).T.squeeze(dim=0)
+        out = tanh_polys_flipped.matmul(coeffs)
+
+        # truncate outside [-maxval, maxval]
+        return hardtanh(out)
```
[the remark this answers is not in the document]
Comment (Syzygianinfern0) [reply src/sympc/approximations/tanh.py:55]: Sometimes, the approximation gives obscurely large (+/-) results. This basically clamps such cases.

File: src/sympc/approximations/tanh.py [hunk 18]
```diff
@@ -31,5 +43,94 @@ def tanh(tensor, method="sigmoid"):
...
+    Args:
+        tensor (MPCTensor): whose hardtanh has to be calculated
+        min_value (float): minimum value of the linear region range. Default: -1
+        max_value (float): maximum value of the linear region range. Default: 1
+
+    Returns:
+        MPCTensor: calculated MPCTensor
+    """
+    intermediate = [relu(tensor - min_value), relu(tensor - max_value)]
+    intermediate = intermediate[0] - intermediate[1]
```
[the remark this answers is not in the document]
Comment (Syzygianinfern0) [reply src/sympc/approximations/tanh.py:79]: Will do!

File: tests/sympc/tensor/static_test.py (hunk 15 above)
Comment (Syzygianinfern0) [reply tests/sympc/tensor/static_test.py:24]: Its just max/argmax, hence the results are accurate 😅

File: src/sympc/approximations/tanh.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (LaRiffle) [reply src/sympc/approximations/tanh.py:55]: How big are they? Like +10 or +3000
When number are too big, the comparison might fail as well, but + - 10 is fine for sure

[the remark this answers is not in the document]
Comment (Syzygianinfern0) [reply src/sympc/approximations/tanh.py:55]: `tanh` is expected to be [-1,1]. Hence, even +-10 is quite a lot.

File: tests/sympc/tensor/static_test.py (hunk 15 above)
Comment (gmuraru) [reply tests/sympc/tensor/static_test.py:24]: max/argmax -> stack ^_^

**Review by aanurraj — APPROVED** [review]
Lots of good stuff in here 🎉

**Comment by gmuraru** [comment]
Merging this PR since it might be a flaky test

**Pull request merged**